subsys/mgmt/osdp/src/osdp_sc.c

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) 2019 Siddharth Chandrasekaran <siddharth@embedjournal.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(osdp, CONFIG_OSDP_LOG_LEVEL);
+
+#include <string.h>
+#include "osdp_common.h"
+
+#define TAG "SC: "
+
+#define OSDP_SC_EOM_MARKER             0x80  /* End of Message Marker */
+
+/* Default key as specified in OSDP specification */
+static const uint8_t osdp_scbk_default[16] = {
+	0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37,
+	0x38, 0x39, 0x3A, 0x3B, 0x3C, 0x3D, 0x3E, 0x3F
+};
+
+void osdp_compute_scbk(struct osdp_pd *pd, uint8_t *scbk)
+{
+	int i;
+	struct osdp *ctx = TO_CTX(pd);
+
+	memcpy(scbk, pd->sc.pd_client_uid, 8);
+	for (i = 8; i < 16; i++) {
+		scbk[i] = ~scbk[i - 8];
+	}
+	osdp_encrypt(ctx->sc_master_key, NULL, scbk, 16);
+}
+
+void osdp_compute_session_keys(struct osdp *ctx)
+{
+	int i;
+	struct osdp_pd *pd = GET_CURRENT_PD(ctx);
+
+	if (ISSET_FLAG(pd, PD_FLAG_SC_USE_SCBKD)) {
+		memcpy(pd->sc.scbk, osdp_scbk_default, 16);
+	} else {
+		/**
+		 * Compute SCBK only in CP mode. PD mode, expect to already have
+		 * the SCBK (sent from application layer).
+		 */
+		if (ISSET_FLAG(pd, PD_FLAG_PD_MODE) == 0) {
+			osdp_compute_scbk(pd, pd->sc.scbk);
+		}
+	}
+
+	memset(pd->sc.s_enc, 0, 16);
+	memset(pd->sc.s_mac1, 0, 16);
+	memset(pd->sc.s_mac2, 0, 16);
+
+	pd->sc.s_enc[0]  = 0x01;  pd->sc.s_enc[1]  = 0x82;
+	pd->sc.s_mac1[0] = 0x01;  pd->sc.s_mac1[1] = 0x01;
+	pd->sc.s_mac2[0] = 0x01;  pd->sc.s_mac2[1] = 0x02;
+
+	for (i = 2; i < 8; i++) {
+		pd->sc.s_enc[i]  = pd->sc.cp_random[i - 2];
+		pd->sc.s_mac1[i] = pd->sc.cp_random[i - 2];
+		pd->sc.s_mac2[i] = pd->sc.cp_random[i - 2];
+	}
+
+	osdp_encrypt(pd->sc.scbk, NULL, pd->sc.s_enc,  16);
+	osdp_encrypt(pd->sc.scbk, NULL, pd->sc.s_mac1, 16);
+	osdp_encrypt(pd->sc.scbk, NULL, pd->sc.s_mac2, 16);
+}
+
+void osdp_compute_cp_cryptogram(struct osdp_pd *pd)
+{
+	/* cp_cryptogram = AES-ECB( pd_random[8] || cp_random[8], s_enc ) */
+	memcpy(pd->sc.cp_cryptogram + 0, pd->sc.pd_random, 8);
+	memcpy(pd->sc.cp_cryptogram + 8, pd->sc.cp_random, 8);
+	osdp_encrypt(pd->sc.s_enc, NULL, pd->sc.cp_cryptogram, 16);
+}
+
+/**
+ * Like memcmp; but operates at constant time.
+ *
+ * Returns:
+ *    0: if memory pointed to by s1 and and s2 are identical.
+ *  +ve: if the they are different.
+ */
+static int osdp_ct_compare(const void *s1, const void *s2, size_t len)
+{
+	size_t i, ret = 0;
+	const uint8_t *_s1 = s1;
+	const uint8_t *_s2 = s2;
+
+	for (i = 0; i < len; i++) {
+		if (_s1[i] != _s2[i]) {
```

Comment:
This isn't actually constant time, as you conditionally increment based on whether the bytes match. You should do something like:
```C
for (i = 0; i < len; i++) {
        ret |= _s1[i] ^ _s2[i];
}
return ret;
```

Reply:
Very nice, thank you.